Source/Core/DiscIO/DirectoryBlob.h

```diff
@@ -0,0 +1,128 @@
+// Copyright 2008 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <set>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "DiscIO/Blob.h"
+
+namespace File
+{
+struct FSTEntry;
+class IOFile;
+}
+
+namespace DiscIO
+{
+class DirectoryBlobReader : public BlobReader
+{
+public:
+  static std::unique_ptr<DirectoryBlobReader> Create(const std::string& dol_path);
+
+  bool Read(u64 offset, u64 length, u8* buffer) override;
+  bool SupportsReadWiiDecrypted() const override;
+  bool ReadWiiDecrypted(u64 offset, u64 size, u8* buffer, u64 partition_offset) override;
+
+  BlobType GetBlobType() const override;
+  u64 GetRawSize() const override;
+  u64 GetDataSize() const override;
+
+private:
+  class DiscContent
+  {
+  public:
+    // Either a pointer to memory or a path to a file can be used as a content source
+    using ContentSource = std::variant<std::string, const u8*>;
+
+    DiscContent(u64 offset, u64 size, ContentSource content_source);
+
+    // Provided because it's convenient when searching for DiscContent in an std::set
+    explicit DiscContent(u64 offset);
+
+    u64 GetOffset() const;
+    u64 GetSize() const;
+    bool Read(u64* offset, u64* length, u8** buffer) const;
+
+    bool operator==(const DiscContent& other) const { return m_offset == other.m_offset; }
+    bool operator!=(const DiscContent& other) const { return !(*this == other); }
+    bool operator<(const DiscContent& other) const { return m_offset < other.m_offset; }
+    bool operator>(const DiscContent& other) const { return other < *this; }
+    bool operator<=(const DiscContent& other) const { return !(*this < other); }
+    bool operator>=(const DiscContent& other) const { return !(*this > other); }
+  private:
+    u64 m_offset;
+    u64 m_size;
+    ContentSource m_content_source;
+  };
+
+  DirectoryBlobReader(const std::string& root_directory);
+
+  bool ReadInternal(u64 offset, u64 length, u8* buffer, const std::set<DiscContent>& contents);
+
+  void SetDiscHeaderAndDiscType();
+  bool SetApploader(const std::string& apploader);
+  void SetDOL();
+
+  void BuildFST();
+
+  void PadToAddress(u64 start_address, u64* address, u64* length, u8** buffer) const;
+
+  void Write32(u32 data, u32 offset, std::vector<u8>* const buffer);
```

Comment:
That's not related to anything I've changed in this PR. GitHub is only showing this line as changed because I've renamed the file.